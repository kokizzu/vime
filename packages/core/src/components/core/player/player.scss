vime-player {
  /**
   * @prop --vime-font-family: A custom font family to be used throughout the player.
   * 
   * @prop --vime-theme: A custom theme (color) to be used throughout the player. Any valid CSS 
   * `color` property should work (HEX, RGBA, HLS, ...).
   */

  --vime-font-family: 'Helvetica Neue', 'Segoe UI', Helvetica, Arial, sans-serif;

  box-sizing: border-box;
  direction: ltr;
  font-family: var(--vime-font-family);
  -moz-osx-font-smoothing: auto;
  -webkit-font-smoothing: subpixel-antialiased;
  -webkit-tap-highlight-color: transparent;
  font-variant-numeric: tabular-nums;
  font-weight: 500;
  line-height: 1.7;
  width: 100%;
  display: block;
  max-width: 100%;
  min-width: 414px;
  position: relative;
  text-shadow: none;
  outline: 0;
  transition: box-shadow 0.3s ease;

  *,
  *::after,
  *::before {
    box-sizing: inherit;
  }

  button {
    font: inherit;
    line-height: inherit;
    width: auto;

    &::-moz-focus-inner {
      border:0;
    }
  }

  // Fix 300ms delay
  a,
  button,
  input,
  label {
    touch-action: manipulation;
  }

  &.video {
    height: 0;
    background: #000;
    overflow: hidden;
  }

  &.fullscreen {
    background: #000;
    margin: 0;
    border-radius: 0;
    width: 100%;
    height: 100%;
    padding-bottom: 0 !important;
  }

  .blocker {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    z-index: $z-index-blocker;
    display: inline-block;
  }
}
