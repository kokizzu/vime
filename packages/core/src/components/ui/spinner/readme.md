# vime-spinner

Displays a loading indicator when the video is `buffering`.

## Example

```html
<vime-player>
  <!-- ... -->
  <vime-ui>
    <vime-spinner></vime-spinner>
  </vime-ui>
</vime-player>
```

<!-- Auto Generated Below -->


## Events

| Event      | Description                              | Type                |
| ---------- | ---------------------------------------- | ------------------- |
| `willHide` | Emitted when the spinner will be hidden. | `CustomEvent<void>` |
| `willShow` | Emitted when the spinner will be shown.  | `CustomEvent<void>` |


## CSS Custom Properties

| Name                         | Description                                                |
| ---------------------------- | ---------------------------------------------------------- |
| `--spinner-fill-color`       | The color of the progress within the track.                |
| `--spinner-height`           | The height of the spinner.                                 |
| `--spinner-spin-duration`    | How long it takes the spinner to complete a full rotation. |
| `--spinner-spin-timing-func` | The animation timing function to use for the spin.         |
| `--spinner-thickness`        | The thickness of the spinner in px.                        |
| `--spinner-track-color`      | The color of the track the spinner is rotating in.         |
| `--spinner-width`            | The width of the spinner.                                  |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
