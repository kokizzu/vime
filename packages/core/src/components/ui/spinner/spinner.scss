vime-spinner {
  /**
   * @prop --width: The width of the spinner.
   * @prop --height: The height of the spinner.
   * @prop --thickness: The thickness of the spinner in px.
   * @prop --track-color: The color of the track the spinner is rotating in.
   * @prop --progress-color: The color of the progress within the track.
   * @prop --spin-duration: How long it takes the spinner to complete a full rotation.
   * @prop --spin-timing-func: The animation timing function to use.
   */

  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  justify-content: center;
  align-items: center;
  pointer-events: none;
  z-index: $z-index-spinner;
  display: none;
  opacity: 0;
  visibility: hidden;
  pointer-events: none;
  transition: $fade-transition;
    
  &.enabled {
    display: flex;
  }

  &.active {
    opacity: 1;
    visibility: visible;
  }
  
  @keyframes spin {
    0% {
      transform: rotate(0deg);
    }
    100% {
      transform: rotate(360deg);
    }
  }

  & > div {
    background: transparent;
    margin: 60px auto;
    font-size: 10px;
    position: relative;
    text-indent: -9999em;
    border-top: var(--thickness, 3px) solid var(--progress-color, #fff);
    border-left: var(--thickness, 3px) solid var(--progress-color, #fff);
    border-right: var(--thickness, 3px) solid var(--track-color, $color-white-200);
    border-bottom: var(--thickness, 3px) solid var(--track-color, $color-white-200);
    transform: translateZ(0);
    animation: spin var(--spin-duration, 1.1s) infinite var(--spin-timing-func, linear);

    &,
    &::after {
      border-radius: 50%;
      width: var(--width, 80px);
      height: var(--height, 80px);
    }
  }
}