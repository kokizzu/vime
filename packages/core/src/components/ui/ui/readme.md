# vime-ui

Simple container that holds a collection of user interface components.

## Example

```html
<vime-player>
  <!-- ... -->

  <vime-ui>
    <vime-poster />
    <!-- ... -->
  </vime-ui>
</vime-player>
```

<!-- Auto Generated Below -->


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
