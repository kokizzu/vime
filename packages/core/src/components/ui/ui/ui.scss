vime-ui {
  position: relative;
  pointer-events: none;
  z-index: $z-index-ui;
  width: 100%;

  &.hidden {
    display: none;
  }

  &.video {
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
  }
}