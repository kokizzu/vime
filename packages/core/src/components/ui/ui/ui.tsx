import { h, Component } from '@stencil/core';

@Component({
  tag: 'vime-ui',
  styleUrl: 'ui.scss',
})
export class UI {
  render() {
    return (<slot />);
  }
}
