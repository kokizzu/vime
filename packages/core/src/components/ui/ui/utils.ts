import { getElement } from '@stencil/core';
import { HTMLStencilElement } from '@stencil/core/internal';

export const findUIRoot = (ref: any) => {
  let ui = getElement(ref);

  while (!(/^VIME-UI$/.test(ui?.nodeName))) {
    ui = ui.parentElement as HTMLStencilElement;
  }

  return ui as HTMLVimeUiElement;
};
