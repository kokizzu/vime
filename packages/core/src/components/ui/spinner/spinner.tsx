import {
  h, Component, State, Prop, Watch, Host,
} from '@stencil/core';
import { PlayerProps, PlayerProp } from '../../core/player/PlayerProp';
import { openPlayerWormhole } from '../../core/player/PlayerWormhole';

@Component({
  tag: 'vime-spinner',
  styleUrl: 'spinner.scss',
})
export class Spinner {
  @State() isEnabled = false;

  @State() isActive = false;

  /**
   * @internal
   */
  @Prop() isVideoView!: PlayerProps[PlayerProp.IsVideoView];

  /**
   * @internal
   */
  @Prop() buffering!: PlayerProps[PlayerProp.Buffering];

  @Watch('isVideoView')
  onEnabledChange() {
    this.isEnabled = this.isVideoView;
  }

  @Watch('buffering')
  onActiveChange() {
    this.isActive = this.buffering;
  }

  render() {
    return (
      <Host
        class={{
          enabled: this.isEnabled,
          active: this.isActive,
        }}
      >
        <div>Loading...</div>
      </Host>
    );
  }
}

openPlayerWormhole(Spinner, [
  PlayerProp.IsVideoView,
  PlayerProp.Buffering,
]);
