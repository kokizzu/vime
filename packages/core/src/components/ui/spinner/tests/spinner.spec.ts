import { SpecPage } from '@stencil/core/testing';
import { Spinner } from '../spinner';
import { newUISpecPage } from '../../ui/tests';
import { PlayerProp } from '../../../core/player/PlayerProp';
import { ViewType } from '../../../core/player/ViewType';

let page: SpecPage;
let provider: HTMLVimeFaketubeElement;
let spinner: HTMLVimeSpinnerElement;

beforeEach(async () => {
  ({ page, provider } = await newUISpecPage(
    [Spinner],
    '<vime-spinner />',
  ));

  spinner = page.root!.querySelector('vime-spinner')!;
});

it('should be structurally sound', () => {
  expect(spinner).toMatchSnapshot();
});

it('should not render if not a video view', async () => {
  await provider.dispatchStateChange(PlayerProp.ViewType, ViewType.Audio);
  await page.waitForChanges();
  expect(spinner).not.toHaveClass('enabled');
});

it('should render if a video view', async () => {
  await provider.dispatchStateChange(PlayerProp.ViewType, ViewType.Video);
  await page.waitForChanges();
  expect(spinner).toHaveClass('enabled');
});

it('should be visible if buffering', async () => {
  await provider.dispatchStateChange(PlayerProp.Buffering, true);
  await page.waitForChanges();
  expect(spinner).toHaveClass('active');
});

it('should not be visible if not buffering', async () => {
  await provider.dispatchStateChange(PlayerProp.Buffering, false);
  await page.waitForChanges();
  expect(spinner).not.toHaveClass('active');
});
