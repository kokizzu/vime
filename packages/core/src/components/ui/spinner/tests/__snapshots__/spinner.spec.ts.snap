// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be structurally sound 1`] = `
<vime-spinner>
  <div>
    Loading...
  </div>
</vime-spinner>
`;
